pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/DockerPipelineScript.groovy

```diff
@@ -39,7 +39,7 @@ class DockerPipelineScript extends AbstractDockerPipelineScript<DockerPipeline>
     @Override
     Closure runImage(Closure body) {
         return {
-            if (!Utils.withinAStage()) {
+            if (!Utils.withinAStage() && describable.alwaysPull) {
```

Comment:
Hrrrrm. So I'm assuming what you want here is the ability to *not* always pull? Because the existing behavior for a top-level `docker` agent is to always pull.

Reply:
Yes. That's what the issue is about. `alwaysPull` is ignored in the top level agent directive.

I think that users often do want to "always pull" and not pulling is mostly used for local/quick testing. So you might want to reconsider the default of `alwaysPull` which I think is `false`.

Reply:
So that makes me want to think on this for a while. Right now the effective behavior is "always explicitly pull for a top-level agent, do not explicitly pull for a stage agent unless `alwaysPull` is set". So if we add `alwaysPull` checking to the top level and turn the default for `alwaysPull` to false, the new effective behavior will be "always pull for either top level or stage agent, unless `alwaysPull` is explicitly set to `false`".

That concerns me a bit - we deliberately set up the current behavior for stage-level agents and I don't think we want to change that. The top-level behavior for `docker` is different than at the stage level for probably no longer relevant historical reasons - we wanted to keep the `docker pull ...` logs inside a stage that Blue Ocean knows to not display unless there's a problem. But we're seeing a lot more stage-level `docker` agent usage than we expected at the time, so whether or not `alwaysPull` is set, any time someone's using a Docker image that isn't already cached locally, the Docker pull logs are showing up in that stage's log anyway. That is a little different than at the top level, because if we get rid of the explicit `docker pull` call, the pull logs will show up when `docker run` is called and won't be in any stage at all.

Hrrrrrm.

Reply:
Also keep in mind that not using `docker pull` might end up using stale images and failing builds that require an up-to-date image. But sometimes it's useful to not `docker pull`, e.g. testing a local image.

Reply:
Ok, let's live a little - merging, but if there are reports of issues, we may revert later. Thanks!